File: pyslam/slam/cpp/tracking_core.cpp
#include "tracking_core.h"

// Utility functions implementation
#include <Eigen/Dense>
#include <opencv2/opencv.hpp>
#include <vector>

#include "Eigen/Dense"
#include "config_parameters.h"
#include "frame.h"
#include "map_point.h"
#include "smart_pointers.h"
#include "utils/cv_ops.h"
#include "utils/geom_2views.h"
#include "utils/image_processing.h"
#include "utils/messages.h"

// Add this helper function in tracking_core.cpp (or as an inline function in a header)
namespace {
// Get the best available RANSAC method
inline int get_ransac_method() {
#if defined(OPENCV_VERSION_MAJOR) && OPENCV_VERSION_MAJOR >= 4
#if defined(USAC_MAGSAC)
    return cv::USAC_MAGSAC;
#else
    return cv::RANSAC;
#endif
#else
    return cv::RANSAC;
#endif
}
} // namespace

namespace pyslam {

// Essential matrix pose estimation
std::tuple<std::vector<int>, std::vector<int>, int>
TrackingCore::estimate_pose_by_fitting_ess_mat(const FramePtr &f_ref, FramePtr &f_cur,
                                               const std::vector<int> &idxs_ref,
                                               const std::vector<int> &idxs_cur) {
    int num_inliers = 0;

    if (idxs_ref.empty() || idxs_cur.empty() || idxs_cur.size() != idxs_ref.size()) {
        MSG_RED_WARN("idxs_ref or idxs_cur is empty or len(idxs_cur) != len(idxs_ref)");
        return std::make_tuple(idxs_ref, idxs_cur, num_inliers);
    }

    try {
        const int ransac_method = get_ransac_method();

        // Extract keypoints manually
        MatNx2d kpsn_ref(idxs_ref.size(), 2);
        MatNx2d kpsn_cur(idxs_cur.size(), 2);
        for (size_t i = 0; i < idxs_ref.size(); ++i) {
            kpsn_ref(i, 0) = f_ref->kpsn(idxs_ref[i], 0);
            kpsn_ref(i, 1) = f_ref->kpsn(idxs_ref[i], 1);

            kpsn_cur(i, 0) = f_cur->kpsn(idxs_cur[i], 0);
            kpsn_cur(i, 1) = f_cur->kpsn(idxs_cur[i], 1);
        }

        // Estimate inter frame camera motion by using found keypoint matches
        // output of the following function is: Trc = [Rrc, trc] with ||trc||=1 where c=cur, r=ref
        // and pr = Trc * pc
        const auto [Mrc, mask_match] = geom_2views::estimate_pose_ess_mat(
            kpsn_ref, kpsn_cur, ransac_method, Parameters::kRansacProb,
            Parameters::kRansacThresholdNormalized);

        // Mcr = inv_T(Mrc)
        const auto Mcr = inv_T(Mrc);

        // estimated_Tcw = Mcr * f_ref.pose()
        const auto f_ref_pose = f_ref->pose();
        const auto estimated_Tcw = Mcr * f_ref_pose;

        // Remove outliers from keypoint matches by using the mask
        std::vector<int> idxs_ref_out, idxs_cur_out;
        const size_t num_matches = mask_match.size();

        for (size_t i = 0; i < num_matches; ++i) {
            if (mask_match[i] == 1) {
                idxs_ref_out.push_back(idxs_ref[i]);
                idxs_cur_out.push_back(idxs_cur[i]);
            }
        }

        num_inliers = static_cast<int>(idxs_ref_out.size());
        MSG_INFO_STREAM("# inliers: " << num_inliers);

        // If there are not enough inliers do not use the estimated pose
        if (num_inliers < Parameters::kNumMinInliersEssentialMat) {
            MSG_RED_WARN("Essential mat: not enough inliers!");
        } else {
            // Use the estimated pose as an initial guess for the subsequent pose optimization
            // Set only the estimated rotation (essential mat computation does not provide a scale
            // for the translation)
            const Mat3d Rcw = estimated_Tcw.block<3, 3>(0, 0); // copy only the rotation
            const Vec3d tcw =
                f_ref_pose.block<3, 1>(0, 3); // override translation with ref frame translation
            f_cur->update_rotation_and_translation(Rcw, tcw);
        }

        return std::make_tuple(std::move(idxs_ref_out), std::move(idxs_cur_out), num_inliers);

    } catch (const std::exception &e) {
        MSG_RED_WARN_STREAM("Error in estimate_pose_ess_mat: " << e.what());
        return std::make_tuple(idxs_ref, idxs_cur, num_inliers);
    }
}

// Homography estimation with RANSAC
std::tuple<bool, std::vector<int>, std::vector<int>, int, int>
TrackingCore::find_homography_with_ransac(const FramePtr &f_cur, const FramePtr &f_ref,
                                          const std::vector<int> &idxs_cur,
                                          const std::vector<int> &idxs_ref,
                                          const double reproj_threshold /* = 5.0 */,
                                          const int min_num_inliers /* = 15 */) {
    int num_matched_kps = 0;
    int num_outliers = 0;

    if (idxs_cur.empty() || idxs_cur.size() != idxs_ref.size()) {
        MSG_RED_WARN(
            "find_homography_with_ransac: idxs_cur is empty or len(idxs_cur) != len(idxs_ref)");
        return std::make_tuple(false, std::vector<int>(), std::vector<int>(), 0, 0);
    }

    const int ransac_method = get_ransac_method();

    // Get keypoints
    std::vector<cv::Point2f> kps_cur, kps_ref;
    kps_cur.reserve(idxs_cur.size());
    kps_ref.reserve(idxs_ref.size());
    for (size_t i = 0; i < idxs_cur.size(); ++i) {
        const float x_cur = f_cur->kps(idxs_cur[i], 0);
        const float y_cur = f_cur->kps(idxs_cur[i], 1);
        const float x_ref = f_ref->kps(idxs_ref[i], 0);
        const float y_ref = f_ref->kps(idxs_ref[i], 1);
        kps_cur.emplace_back(x_cur, y_cur);
        kps_ref.emplace_back(x_ref, y_ref);
    }

    cv::Mat H, mask;
    H = cv::findHomography(kps_cur, kps_ref, ransac_method, reproj_threshold, mask);

    // Check if homography was found (mask is not None)
    if (mask.empty()) {
        return std::make_tuple(false, std::vector<int>(), std::vector<int>(), 0, 0);
    }

    // Convert mask to boolean and count inliers
    int num_inliers = 0;
    std::vector<int> idxs_cur_out, idxs_ref_out;
    idxs_cur_out.reserve(idxs_cur.size());
    idxs_ref_out.reserve(idxs_ref.size());

    for (int i = 0; i < mask.rows; ++i) {
        if (mask.at<uchar>(i) == 1) {
            num_inliers++;
            idxs_cur_out.push_back(idxs_cur[i]);
            idxs_ref_out.push_back(idxs_ref[i]);
        }
    }

    num_outliers = static_cast<int>(idxs_cur.size()) - num_inliers;

    if (num_inliers < min_num_inliers) {
        return std::make_tuple(false, std::vector<int>(), std::vector<int>(), 0, 0);
    }

    num_matched_kps = static_cast<int>(idxs_cur_out.size());
    return std::make_tuple(true, std::move(idxs_cur_out), std::move(idxs_ref_out), num_matched_kps,
                           num_outliers);
}

// propagate map point matches from f_ref to f_cur (access frames from tracking thread, no need to
// lock)
std::tuple<int, std::vector<int>, std::vector<int>> TrackingCore::propagate_map_point_matches(
    const FramePtr &f_ref, FramePtr &f_cur, const std::vector<int> &idxs_ref,
    const std::vector<int> &idxs_cur, float max_descriptor_distance /* = -1.0f*/) {

    if (max_descriptor_distance < 0) {
        max_descriptor_distance = Parameters::kMaxDescriptorDistance; // take the updated value
    }

    int num_matched_map_pts = 0;
    std::vector<int> idx_ref_out, idx_cur_out;
    idx_ref_out.reserve(idxs_ref.size());
    idx_cur_out.reserve(idxs_cur.size());

    // populate f_cur with map points by propagating map point matches of f_ref;
    // to this aim, we use map points observed in f_ref and keypoint matches between f_ref and f_cur
    const auto &pts_ref = f_ref->points;
    const auto &outliers_ref = f_ref->outliers;
    const auto &pts_cur = f_cur->points;
    const auto &des_cur = f_cur->des;

    const int idxs_ref_size = static_cast<int>(idxs_ref.size());
    const int idxs_cur_size = static_cast<int>(idxs_cur.size());
    const int pts_ref_size = static_cast<int>(f_ref->points.size());
    const int pts_cur_size = static_cast<int>(f_cur->points.size());

    for (int i = 0; i < idxs_ref_size; ++i) {
        const int idx_ref = idxs_ref[i];
        if (idx_ref < 0 || idx_ref >= pts_ref_size)
            continue;
        const MapPointPtr &p_ref = pts_ref[idx_ref];
        if (!p_ref || outliers_ref[idx_ref] || p_ref->is_bad())
            continue;

        const int idx_cur = idxs_cur[i];
        if (idx_cur < 0 || idx_cur >= pts_cur_size)
            continue;
        const MapPointPtr &p_cur = pts_cur[idx_cur];
        if (p_cur)
            continue; // already has a match

        // distance to descriptor
        float des_dist = p_ref->min_des_distance(des_cur.row(idx_cur));
        if (des_dist > max_descriptor_distance)
            continue;

        if (p_ref->add_frame_view(f_cur, idx_cur)) {
            // => P is matched to the i-th matched keypoint in f_cur
            ++num_matched_map_pts;
            idx_ref_out.push_back(idx_ref);
            idx_cur_out.push_back(idx_cur);
        }
    }
    return std::make_tuple(num_matched_map_pts, std::move(idx_ref_out), std::move(idx_cur_out));
}

std::vector<MapPointPtr> TrackingCore::create_vo_points(FramePtr &frame, int max_num_points,
                                                        const Vec3b &color) {
    std::lock_guard<std::mutex> lock(frame->_lock_features);

    const auto &depths = frame->depths;
    const auto &camera = frame->camera;
    auto &points = frame->points;

    // Check depths are available
    if (depths.empty()) {
        MSG_WARN("TrackingCore::create_vo_points() - depths are empty");
        return {};
    }

    // Use camera depth threshold if not provided
    const float depth_threshold = camera->depth_threshold;

    // Collect all points with valid depths
    std::vector<std::pair<int, float>> valid_idxs_depths;
    valid_idxs_depths.reserve(depths.size());
    for (int i = 0; i < static_cast<int>(depths.size()); ++i) {
        if (depths[i] > Parameters::kMinDepth) {
            valid_idxs_depths.push_back(std::make_pair(i, depths[i]));
        }
    }
    if (valid_idxs_depths.empty()) {
        return {};
    }

    // Sort by depth (increasing order)
    std::sort(valid_idxs_depths.begin(), valid_idxs_depths.end(),
              [](const std::pair<int, float> &a, const std::pair<int, float> &b) {
                  return a.second < b.second;
              });

    int num_vo_points = 0;
    std::vector<MapPointPtr> vo_points;
    vo_points.reserve(max_num_points);

    // Insert all points that are within the depth threshold or at least the closest max_num_points
    // points
    for (const auto &[idx, depth] : valid_idxs_depths) {
        const auto &p = points[idx];
        // If there is not a point or the point is not already observed,
        // then create a new map point
        if (!p || (p && p->num_observations() < 1)) {
            auto [pt3d, is_valid] =
                frame->unproject_point_3d<double>(idx, /*transform_in_world*/ true);
            if (is_valid) {
                auto mp = std::make_shared<MapPoint>(pt3d, color, frame, idx);
                vo_points.push_back(mp);
                points[idx] = mp;
                num_vo_points++;
            }
        }

        // Stop when we have selected at least a maximum number of points and the depth is greater
        // than the threshold
        if (num_vo_points >= max_num_points && depth > depth_threshold) {
            break;
        }
    }
    return vo_points;
}

int TrackingCore::create_and_add_stereo_map_points_on_new_kf(FramePtr &frame, KeyFramePtr &kf,
                                                             MapPtr &map, const cv::Mat &img) {

    const auto &depths = kf->depths;
    const auto &camera = kf->camera;
    auto &frame_points = frame->points;
    auto &kf_points = kf->points;

    // Check depths are available
    if (depths.empty()) {
        MSG_WARN("TrackingCore::create_and_add_stereo_map_points_on_new_kf() - depths are empty");
        return {};
    }

    // Use camera depth threshold if not provided
    const float depth_threshold = camera->depth_threshold;

    // Collect all points with valid depths
    std::vector<std::pair<int, float>> valid_idxs_depths;
    valid_idxs_depths.reserve(depths.size());
    for (int i = 0; i < static_cast<int>(depths.size()); ++i) {
        if (depths[i] > Parameters::kMinDepth) {
            valid_idxs_depths.push_back(std::make_pair(i, depths[i]));
        }
    }
    if (valid_idxs_depths.empty()) {
        return 0;
    }

    // Sort by depth (increasing order)
    std::sort(valid_idxs_depths.begin(), valid_idxs_depths.end(),
              [](const std::pair<int, float> &a, const std::pair<int, float> &b) {
                  return a.second < b.second;
              });

    const int max_num_stereo_points = Parameters::kMaxNumStereoPointsOnNewKeyframe;
    int num_stereo_points = 0;

    utils::ImageColorExtractor img_color_extractor(img);
    const int num_channels = img.channels();
    if (num_channels != 1 && num_channels != 3) {
        MSG_ERROR("TrackingCore::create_and_add_stereo_map_points_on_new_kf() - image channels "
                  "are not 1 or 3");
        return 0;
    }
    const bool is_color = num_channels == 3;

    // Insert all points that are within the depth threshold or at least the closest max_num_points
    // points
    for (const auto &[idx, depth] : valid_idxs_depths) {
        const auto &p = kf_points[idx];
        // If there is not a point or the point is not already observed,
        // then create a new map point
        if (!p || (p && p->num_observations() < 1)) {
            auto [pt3d, is_valid] =
                frame->unproject_point_3d<double>(idx, /*transform_in_world*/ true);
            if (is_valid) {
                const auto kp = frame->kps.row(idx);
                const auto color_cv = img_color_extractor.extract_mean_color(kp, is_color);
                // Convert cv::Vec<float,3> to Eigen::Matrix<unsigned char,3,1>
                Vec3b color_eigen(
                    static_cast<unsigned char>(std::clamp(color_cv[0], 0.0f, 255.0f)),
                    static_cast<unsigned char>(std::clamp(color_cv[1], 0.0f, 255.0f)),
                    static_cast<unsigned char>(std::clamp(color_cv[2], 0.0f, 255.0f)));
                auto mp = std::make_shared<MapPoint>(pt3d, color_eigen, kf, idx);

                frame_points[idx] = mp; // add point to the frame
                map->add_point(mp);     // add point to the map
                mp->add_observation(kf, idx);
                mp->update_info();
                num_stereo_points++;
            }
        }

        // Stop when we have selected at least a maximum number of points and the depth is greater
        // than the threshold
        if (num_stereo_points >= max_num_stereo_points && depth > depth_threshold) {
            break;
        }
    }
    return num_stereo_points;
}

std::tuple<int, int, std::vector<bool>>
TrackingCore::count_tracked_and_non_tracked_close_points(const FramePtr &f_cur,
                                                         const SensorType sensor_type) {
    // Count how many "close" points are being tracked and how many "close" points could be
    // potentially created. For convenience, we also compute and return a mask of tracked points.

    int num_non_tracked_close = 0;
    int num_tracked_close = 0;
    std::vector<bool> tracked_mask;

    const auto &points = f_cur->points;
    const auto &outliers = f_cur->outliers;
    const auto &depths = f_cur->depths;
    const auto &camera = f_cur->camera;
    const float depth_threshold = camera->depth_threshold;
    const size_t num_points = points.size();

    if (sensor_type != SensorType::MONOCULAR) {
        tracked_mask.resize(num_points, false);
        for (size_t i = 0; i < num_points; ++i) {
            const bool is_close =
                (depths[i] > Parameters::kMinDepth && depths[i] < depth_threshold);
            const bool is_tracked = points[i] && !outliers[i];
            if (is_tracked) { // point is tracked
                tracked_mask[i] = true;
                if (is_close) { // point is close and tracked
                    num_tracked_close++;
                }
            } else if (is_close) { // point is close but not tracked
                num_non_tracked_close++;
            }
        }
    }
    return std::make_tuple(num_tracked_close, num_non_tracked_close, std::move(tracked_mask));
}

} // namespace pyslam

File: pyslam/slam/tracking_core.py
"""
* This file is part of PYSLAM
*
* Copyright (C) 2016-present Luigi Freda <luigi dot freda at gmail dot com>
*
* PYSLAM is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* PYSLAM is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with PYSLAM. If not, see <http://www.gnu.org/licenses/>.
"""

import numpy as np
import cv2

# Import MapPoint locally to avoid circular imports
from pyslam.config_parameters import Parameters
from pyslam.slam.map_point import MapPoint
from pyslam.slam.frame import Frame
from pyslam.slam.keyframe import KeyFrame
from pyslam.slam.map import Map
from pyslam.io.dataset_types import SensorType
from pyslam.utilities.logging import Printer
from pyslam.utilities.geom_2views import estimate_pose_ess_mat
from pyslam.utilities.geometry import inv_T

from typing import Tuple

kRansacThresholdNormalized = (
    0.0004  # 0.0003 # metric threshold used for normalized image coordinates
)
kRansacProb = 0.999
kNumMinInliersEssentialMat = 8

kRansacReprojThreshold = 5
kRansacMinNumInliers = 15


class TrackingCore:

    # estimate a pose from a fitted essential mat;
    # since we do not have an interframe translation scale, this fitting can be used to detect outliers, estimate interframe orientation and translation direction
    # N.B. read the NBs of the method estimate_pose_ess_mat(), where the limitations of this method are explained
    @staticmethod
    def estimate_pose_by_fitting_ess_mat(
        f_ref: Frame, f_cur: Frame, idxs_ref: list[int], idxs_cur: list[int]
    ) -> Tuple[list[int], list[int], int]:
        num_inliers = 0
        mask_match = None

        if len(idxs_ref) == 0 or len(idxs_cur) != len(idxs_ref):
            Printer.red(f"idxs_ref or idxs_cur is empty or len(idxs_cur) != len(idxs_ref)")
            # mask_match = np.zeros(len(idxs_ref), dtype=bool)
            return idxs_ref, idxs_cur, num_inliers

        # N.B.: in order to understand the limitations of fitting an essential mat, read the comments of the method self.estimate_pose_ess_mat()
        ransac_method = None
        try:
            ransac_method = cv2.USAC_MAGSAC
        except:
            ransac_method = cv2.RANSAC
        try:
            # estimate inter frame camera motion by using found keypoint matches
            # output of the following function is:  Trc = [Rrc, trc] with ||trc||=1  where c=cur, r=ref  and  pr = Trc * pc
            Mrc, mask_match = estimate_pose_ess_mat(
                f_ref.kpsn[idxs_ref],
                f_cur.kpsn[idxs_cur],
                method=ransac_method,
                prob=kRansacProb,
                threshold=kRansacThresholdNormalized,
            )
        except Exception as e:
            Printer.red(f"Error in estimate_pose_ess_mat: {e}")
            return idxs_ref, idxs_cur, num_inliers

        Mcr = inv_T(Mrc)
        estimated_Tcw = np.dot(Mcr, f_ref.pose())

        # remove outliers from keypoint matches by using the mask computed with inter frame pose estimation
        mask_idxs = mask_match.ravel() == 1
        num_inliers = sum(mask_idxs)
        print("# inliers: ", num_inliers)
        idxs_ref = idxs_ref[mask_idxs]
        idxs_cur = idxs_cur[mask_idxs]

        # if there are not enough inliers do not use the estimated pose
        if num_inliers < kNumMinInliersEssentialMat:
            # f_cur.update_pose(f_ref.pose) # reset estimated pose to previous frame
            Printer.red("Essential mat: not enough inliers!")
        else:
            # use the estimated pose as an initial guess for the subsequent pose optimization
            # set only the estimated rotation (essential mat computation does not provide a scale for the translation, see above)
            Rcw = estimated_Tcw[:3, :3]  # copy only the rotation
            tcw = f_ref.pose()[:3, 3]  # override translation with ref frame translation
            f_cur.update_rotation_and_translation(Rcw, tcw)
        return idxs_ref, idxs_cur, num_inliers

    # Use a general homography RANSAC matcher with a large threshold of 5 pixels to model the inter-frame transformation for a generic motion
    # NOTE: this method is used to find inliers and estimate the inter-frame transformation (assuming frames are very close in space)
    @staticmethod
    def find_homography_with_ransac(
        f_cur: Frame,
        f_ref: Frame,
        idxs_cur: list[int],
        idxs_ref: list[int],
        reproj_threshold=kRansacReprojThreshold,
        min_num_inliers=kRansacMinNumInliers,
    ) -> Tuple[bool, list[int], list[int], int, int]:
        num_matched_kps = 0
        num_outliers = 0

        if len(idxs_cur) == 0 or len(idxs_cur) != len(idxs_ref):
            Printer.red(
                f"find_homography_with_ransac: idxs_cur is empty or len(idxs_cur) != len(idxs_ref)"
            )
            return False, np.array([], dtype=int), np.array([], dtype=int), 0, 0

        ransac_method = None
        try:
            ransac_method = cv2.USAC_MAGSAC
        except:
            ransac_method = cv2.RANSAC
        kps_cur = f_cur.kps[idxs_cur]
        kps_ref = f_ref.kps[idxs_ref]
        H, mask = cv2.findHomography(
            kps_cur, kps_ref, ransac_method, ransacReprojThreshold=reproj_threshold
        )

        # Check if homography was found (mask is not None)
        if mask is None:
            return False, np.array([], dtype=int), np.array([], dtype=int), 0, 0

        # Ensure mask is a numpy array and convert to boolean
        mask = np.asarray(mask, dtype=bool).ravel()

        num_inliers = np.count_nonzero(mask)
        num_outliers = len(idxs_cur) - num_inliers
        if num_inliers < min_num_inliers:
            return False, np.array([], dtype=int), np.array([], dtype=int), 0, 0
        else:
            # Ensure idxs_cur and idxs_ref are numpy arrays for proper indexing
            idxs_cur = np.asarray(idxs_cur)
            idxs_ref = np.asarray(idxs_ref)

            idxs_cur = idxs_cur[mask]
            idxs_ref = idxs_ref[mask]
            num_matched_kps = len(idxs_cur)
            return True, idxs_cur, idxs_ref, num_matched_kps, num_outliers

    # propagate map point matches from f_ref to f_cur (access frames from tracking thread, no need to lock)
    @staticmethod
    def propagate_map_point_matches(f_ref, f_cur, idxs_ref, idxs_cur, max_descriptor_distance=None):
        # check_orientation=False):
        if max_descriptor_distance is None:
            max_descriptor_distance = Parameters.kMaxDescriptorDistance

        idx_ref_out = []
        idx_cur_out = []

        # rot_histo = RotationHistogram()
        # check_orientation = check_orientation and kCheckFeaturesOrientation and FeatureTrackerShared.oriented_features

        # populate f_cur with map points by propagating map point matches of f_ref;
        # to this aim, we use map points observed in f_ref and keypoint matches between f_ref and f_cur
        num_matched_map_pts = 0
        for i, idx_ref in enumerate(idxs_ref):  # iterate over keypoint matches
            p_ref = f_ref.points[idx_ref]
            if (
                p_ref is None or f_ref.outliers[idx_ref] or p_ref.is_bad()
            ):  # do not consider pose optimization outliers or bad points
                continue
            idx_cur = idxs_cur[i]
            p_cur = f_cur.points[idx_cur]
            if (
                p_cur is not None
            ):  # and p_cur.num_observations() > 0: # if we already matched p_cur => no need to propagate anything
                continue
            des_distance = p_ref.min_des_distance(f_cur.des[idx_cur])
            if des_distance > max_descriptor_distance:
                continue
            if p_ref.add_frame_view(
                f_cur, idx_cur
            ):  # => P is matched to the i-th matched keypoint in f_cur
                num_matched_map_pts += 1
                idx_ref_out.append(idx_ref)
                idx_cur_out.append(idx_cur)

                # if check_orientation:
                #     index_match = len(idx_cur_out)-1
                #     rot = f_ref.angles[idx]-f_cur.angles[idx_cur]
                #     rot_histo.push(rot, index_match)

        # if check_orientation:
        #     valid_match_idxs = rot_histo.get_valid_idxs()
        #     print('checking orientation consistency - valid matches % :', len(valid_match_idxs)/max(1,len(idxs_cur))*100,'% of ', len(idxs_cur),'matches')
        #     #print('rotation histogram: ', rot_histo)
        #     idx_ref_out = np.array(idx_ref_out)[valid_match_idxs]
        #     idx_cur_out = np.array(idx_cur_out)[valid_match_idxs]
        #     num_matched_map_pts = len(valid_match_idxs)

        return num_matched_map_pts, idx_ref_out, idx_cur_out

    @staticmethod
    def create_vo_points(
        frame: Frame, max_num_points=Parameters.kMaxNumVisualOdometryPoints, color=(0, 0, 255)
    ):
        """
        Create VO (Visual Odometry) points on this frame using depth information.

        This method selects keypoints with valid depths, filters them based on depth thresholds
        and observation counts, then creates MapPoint objects for tracking purposes.

        Args:
            max_num_points: Maximum number of VO points to create
            depth_threshold: Depth threshold for point selection (uses camera.depth_threshold if None)
            color: Color for VO points (BGR format)

        Returns:
            list: List of created MapPoint objects
        """

        if frame.depths is None:
            return []

        depth_threshold = frame.camera.depth_threshold

        # Filter points with valid depths
        valid_mask = frame.depths > Parameters.kMinDepth
        if not np.any(valid_mask):
            return []

        valid_depths = frame.depths[valid_mask]
        valid_indices = np.where(valid_mask)[0]

        # Sort by depth (increasing order)
        sort_indices = np.argsort(valid_depths)
        sorted_depths = valid_depths[sort_indices]
        sorted_indices = valid_indices[sort_indices]

        # Apply depth threshold and max points selection
        mask_depths_smaller_than_th = sorted_depths < depth_threshold
        mask_first_N_points = np.arange(len(sorted_depths)) < max_num_points
        mask_first_selection = np.logical_or(mask_depths_smaller_than_th, mask_first_N_points)

        # Apply first selection mask
        selected_indices = sorted_indices[mask_first_selection]

        # Filter out points that already have observations
        vector_num_mp_observations = np.array(
            [p.num_observations() if p is not None else 0 for p in frame.points[selected_indices]],
            dtype=np.int32,
        )
        mask_where_to_create_new_map_points = vector_num_mp_observations < 1

        # Apply final selection mask
        final_indices = selected_indices[mask_where_to_create_new_map_points]

        if len(final_indices) == 0:
            return []

        # Create 3D points
        pts3d, pts3d_mask = frame.unproject_points_3d(final_indices, transform_in_world=True)
        if pts3d is None or pts3d_mask is None:
            return []

        # Create MapPoint objects
        created_points = []
        for i, (p, is_valid) in enumerate(zip(pts3d, pts3d_mask)):
            if not is_valid:
                continue

            # Create new map point (VO point)
            mp = MapPoint(p[0:3], color, frame, final_indices[i])
            frame.points[final_indices[i]] = mp
            created_points.append(mp)

        return created_points

    @staticmethod
    def create_and_add_stereo_map_points_on_new_kf(f: Frame, kf: KeyFrame, map: Map, img):
        valid_depths_and_idxs = [
            (z, i) for i, z in enumerate(kf.depths) if z > Parameters.kMinDepth
        ]
        valid_depths_and_idxs.sort()  # increasing-depth order

        if len(valid_depths_and_idxs) == 0:
            Printer.yellow(
                "[create_and_add_stereo_map_points_on_new_kf] no valid depths and idxs found, returning"
            )
            return 0

        sorted_z_values, sorted_idx_values = zip(
            *valid_depths_and_idxs
        )  # unpack the sorted z values and i values into separate lists
        sorted_z_values = np.array(sorted_z_values, dtype=np.float32)
        sorted_idx_values = np.array(sorted_idx_values, dtype=np.int32)

        N = Parameters.kMaxNumStereoPointsOnNewKeyframe
        # create new map points where the depth is smaller than the prefixed depth threshold
        #        or at least N new points with the closest depths
        mask_depths_smaller_than_th = sorted_z_values < kf.camera.depth_threshold
        mask_first_N_points = np.zeros(len(sorted_z_values), dtype=bool)
        mask_first_N_points[: min(N, len(sorted_z_values))] = (
            True  # set True for the first N points otherwise set all True if len(sorted_z_values) < N
        )
        mask_first_selection = np.logical_or(mask_depths_smaller_than_th, mask_first_N_points)

        sorted_z_values = sorted_z_values[mask_first_selection]
        sorted_idx_values = sorted_idx_values[mask_first_selection]
        sorted_points = kf.points[sorted_idx_values]

        # get the points that are None or where the num of observations is smaller than 1
        vector_num_mp_observations = np.array(
            [p.num_observations() if p is not None else 0 for p in sorted_points],
            dtype=np.int32,
        )
        mask_where_to_create_new_map_points = vector_num_mp_observations < 1

        sorted_z_values = sorted_z_values[mask_where_to_create_new_map_points]
        sorted_idx_values = sorted_idx_values[mask_where_to_create_new_map_points]

        pts3d, pts3d_mask = f.unproject_points_3d(sorted_idx_values, transform_in_world=True)
        num_added_points = map.add_stereo_points(pts3d, pts3d_mask, f, kf, sorted_idx_values, img)
        return num_added_points

    @staticmethod
    def count_tracked_and_non_tracked_close_points(f_cur: Frame, sensor_type: SensorType):
        # Count how many "close" points are being tracked and how many "close" points could be potentially created.
        num_non_tracked_close = 0
        num_tracked_close = 0
        tracked_mask = None
        if sensor_type != SensorType.MONOCULAR:
            # Create a mask for tracked points (not None and not an outlier)
            points_mask = np.array([p is not None for p in f_cur.points])
            outliers_mask = ~np.array(f_cur.outliers)
            tracked_mask = points_mask & outliers_mask

            # Create a mask to identify valid depth values within the threshold
            depth_mask = (f_cur.depths > Parameters.kMinDepth) & (
                f_cur.depths < f_cur.camera.depth_threshold
            )
            # Create a mask for tracked points (not None and not an outlier)
            # tracked_mask = (f_cur.points != None) & (~f_cur.outliers)
            # Count points that are close and tracked
            num_tracked_close = np.sum(depth_mask & tracked_mask)
            # Count points that are close but not tracked
            num_non_tracked_close = np.sum(depth_mask & ~tracked_mask)
        return num_tracked_close, num_non_tracked_close, tracked_mask
